hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/HoodieFileIndex.scala

```diff
@@ -340,9 +347,18 @@ case class HoodieFileIndex(spark: SparkSession,
     //       and candidate files are obtained from these file slices.
 
     lazy val queryReferencedColumns = collectReferencedColumns(spark, queryFilters, schema)
-
+    // bucket query index
+    var bucketIds = Option.empty[BitSet]
+    if (bucketIndex.isIndexAvailable && isDataSkippingEnabled) {
+      bucketIds = bucketIndex.filterQueriesWithBucketHashField(queryFilters)
+    }
+    // record index
     lazy val (_, recordKeys) = recordLevelIndex.filterQueriesWithRecordKey(queryFilters)
-    if (!isMetadataTableEnabled || !isDataSkippingEnabled) {
+
+    // index chose
+    if (bucketIndex.isIndexAvailable && bucketIds.isDefined && bucketIds.get.cardinality() > 0) {
+      Option.apply(bucketIndex.getCandidateFiles(allBaseFiles, bucketIds.get))
```

Comment:
Let's switch this condition to `if simple bucket index is enabled`, and by default this optimization is always enabled for bucket index table.

Reply:
> Let's switch this condition to `if simple bucket index is enabled`, and by default this optimization is always enabled for bucket index table.

if bucketIndex can not filter any file, I think we need try other index

Reply:
> And the bucket index file skipping should be orthoganal to the other skipping strategies, let's just try the bucket skipping first then continue with the other strategies.

good idea, but I think if bucket query index can work, it can achieve a good effect itself.

there are a few considerations here. The record-level index itself serves as the primary key index, and combined with bucket indexing, I think there may not be a better effect. Another consideration is column indexing. This index itself has certain requirements for data layout to achieve good results. However, for bucket tables, sorting the layout is not feasible. Therefore, I think combining it with bucket indexing will not yield very favorable results.

Reply:
We are just doing two level of pruning/skipping here:

1. file group skipping with bucket index; (so that the overall candicates was pruned before next step)
2. file skipping within a file group

These two steps should be othogonal and we could have both, maybe RLI does not make sense when hash keys equals primary keys, but when hash keys are sub-set of record keys, we can still have the gains.

And if there are some other predicates like max/min from the column stats, we can even skip a very special file then.

Reply:
> file skipping within a file group
sorry, I can't understand the meaning of this sentence, can you explain it in detail?
in my opinion, a file group consists of multiple file silce. And allBaseeFile will returen latest file slice.

Reply:
If the query specifies max/min predicates, we can still skip this file after we do bucket pruning.

Reply:
still think the optimization should be orthogonal to the others.